src/LinuxTracing/LibunwindstackMultipleOfflineAndProcessMemory.h

```diff
@@ -0,0 +1,105 @@
+            absl::bit_cast<const uint8_t*>(stack_slice_view.data()),
+            stack_slice_view.start_address(), stack_slice_view.end_address())} {};
+  size_t Read(uint64_t addr, void* dst, size_t size) override {
+    return memory_->Read(addr, dst, size);
+  }
+
+  [[nodiscard]] virtual uint64_t start_address() const { return stack_slice_view_.start_address(); }
+  [[nodiscard]] virtual uint64_t end_address() const { return stack_slice_view_.end_address(); }
+  [[nodiscard]] virtual uint64_t size() const { return stack_slice_view_.size(); }
+
+ private:
+  StackSliceView stack_slice_view_;
+  std::shared_ptr<unwindstack::Memory> memory_;
+};
+
+// This custom implementation of `unwindstack::Memory` carries multiple stack slices, each same as
+// `CreateOfflineMemory` would. When requesting to read an address range, the class will go through
+// the stack slices and if one slice fully contains the requested address range it will read from
+// this stack slice. When requesting an address range outside any of the stack samples,
+// the class falls back to reading from the memory of the process online, as `CreateProcessMemory`
+// would.
+// If the process memory is not specified, the fall back to reading process memory will not be done.
+// Having multiple stack slices allows unwinding callstacks that have multiple stacks involved, such
+// as in the case of Wine system calls.
+// The process memory allows unwinding callstacks that involve virtual modules, such as vDSO.
+class LibunwindstackMultipleOfflineAndProcessMemory : public unwindstack::Memory {
+ public:
+  size_t Read(uint64_t addr, void* dst, size_t size) override;
+
+  LibunwindstackMultipleOfflineAndProcessMemory(
+      std::shared_ptr<Memory> process_memory,
+      std::vector<LibunwindstackOfflineMemory> stack_memory_slices)
+      : process_memory_{std::move(process_memory)},
+        stack_memory_slices_{std::move(std::move(stack_memory_slices))} {}
+
+  static std::shared_ptr<Memory> Create(pid_t pid,
+                                        const std::vector<StackSliceView>& stack_slices) {
+    std::vector<LibunwindstackOfflineMemory> stack_memory_slices =
+        CreateOfflineMemorySlices(stack_slices);
+    return std::make_shared<LibunwindstackMultipleOfflineAndProcessMemory>(
+        unwindstack::Memory::CreateProcessMemoryCached(pid), std::move(stack_memory_slices));
+  }
+
+  static std::shared_ptr<Memory> Create(const std::vector<StackSliceView>& stack_slices) {
+    std::vector<LibunwindstackOfflineMemory> stack_memory_slices =
+        CreateOfflineMemorySlices(stack_slices);
+    return std::make_shared<LibunwindstackMultipleOfflineAndProcessMemory>(
+        nullptr, std::move(stack_memory_slices));
+  }
+
+ private:
+  static std::vector<LibunwindstackOfflineMemory> CreateOfflineMemorySlices(
+      const std::vector<StackSliceView>& stack_slices);
+
+  std::shared_ptr<Memory> process_memory_;
+  std::vector<LibunwindstackOfflineMemory> stack_memory_slices_;
```

Comment:
The type of the elements is ultimately and `undinwstack::Memory`, so I think that the name of the variable should end with `memory`. You can probably drop slices and say `stack_memories_`. This also makes a parallel with `process_memory_`.

Reply:
A similar comment applies to all other variables called `stack_memory_slices`.